### DISCUSSION TITLE
Extending and interacting with the Arduino IDE

### DISCUSSION THREAD

**Discussion by earlephilhower** [dankeboy36/vscode-arduino-api#8] [General]
This is only a suggestion, and maybe is better handled in the `arduino-ide` repo, (and maybe already exists)  but I think a few additions could improve the quality of life for users of 3rd party extensions:

* Some accessor to allow extending the pre-existing `Tools` menu for plug-ins.  Basically replicating in some shape or form what IDE 1.x had (nice to have, but not critical)
* A way to write to the `Output` console vs. making a new one (nice to have, but not critical)
* A way to tell the `Serial Monitor` to release/re-acquire the serial port connection.  This is needed for unimpeded FS uploads (IMHO a big quality-of-life improvement)

**Comment by per1234** [comment]
> Some accessor to allow extending the pre-existing Tools menu for plug-ins.

This request is being tracked at https://github.com/arduino/arduino-ide/issues/2135

**Comment by dankeboy36** [comment]
Thanks for the feedback!

> * A way to write to the `Output` console vs. making a new one (nice to have, but not critical)

It's already possible, but I would neither use nor recommend it. If there are other extensions that want to write the shared _Output_, and IDE2 also writes it, there will be chaos. It's better when every extension writes its own output. We can propose a feature to IDE2 to show other channels in the primary UI.

I highly recommend looking into [VS Code tasks](https://code.visualstudio.com/docs/editor/tasks) and the [task provides](https://code.visualstudio.com/api/extension-guides/task-provider). It has a terminal; you do not have to write to an output channel. The Code guys have an example for everything:
 - https://github.com/microsoft/vscode-extension-samples/tree/main/task-provider-sample,
 - https://github.com/microsoft/vscode-extension-samples/tree/main/extension-terminal-sample

In the long run, I want to enable all Arduino IDE 2.x features in VS Code; it will take some time, but if you do not make a tight connection in your extension to the _Output_ channel (or anything too IDE2 specific) now, later, your extension will run in VS Code without zero code changes.

You need to create an [`OutputChannel`](https://code.visualstudio.com/api/references/vscode-api#OutputChannel) with the `'Arduino'` name.

```ts
const channel = vscode.window.createOutputChannel('Arduino');
channel.show();
channel.appendLine('printed from an extension');
```

https://github.com/dankeboy36/vscode-arduino-api/assets/111981763/cfdb29ef-5382-4e8a-8e1a-3f1e707c4147



> * A way to tell the `Serial Monitor` to release/re-acquire the serial port connection.

I do not think IDE2 will expose such a feature ever. Giving extension developers the freedom to open and close connections is a no-go. What should the IDE2 devs do when one of the extensions has a bug and does not release/acquire the monitor connection?
It's too risky. However, I can imagine such an API:

`api.ts`:

```ts
async function runTask<T>(port: Port, fqbn: string, task: () => Promise<T>, timeout?: number): Promise<T>;
```

`extension.ts`:
```ts
await runTask({ protocol: 'serial', address: 'COM2' }, 'rp2040:rp2040:arduino_nano_connect', async () => {
  // by the time your task runs, IDE2 has made sure to disconnect `serial://COM2` monitor for the board if it was running
  
  // do your extension logic, such as FS upload

  // when IDE2 has executed your task, it will make sure to reconnect the monitor if it was running before your task has started to execute.
})
```

Please bear with us with the API evolution. Once something is available for extension devs, IDE2 must support it (forever). So we must be very careful; this is the main reason all APIs are read-only for now, but I am open to any feedback and proposal.

**Reply by earlephilhower** [reply]
Thanks for the pointers on `Task`s and `OutputChannels`.  I've been reading up on `Pseudoterminal`s as well and just picking the right class abstraction is giving me a headache.  There's a lot to digest for someone who's not been doing **both** Node.js and VS Code internals work!

While in general I understand the desire not to dump stuff into the standard Arduino Output frame, in the specific use case I'm doing the IDE really has to be quiescent anyway (and in the 1.x it never seemed to be an issue).  For other extensions which are chatty and long lived, I definitely understand the issue.

I'll definitely keep trying to get a private output terminal since then, like you said, the extension will work w/plain VSCode (i.e. Platform.IO).  But having a simple fallback option is always helpful!

**Reply by earlephilhower** [reply]
FWIW, in the end I really only needed the Pseudoterminal to get a new window to dump some output into.  It may not be very VS-Code-Extension-like (I can write PERL in any language!) but it gets the job done, asynchronously, using `spawn` and some `await`s to grab output as it arrives.

Thx again for the help and pointers!
